### ISSUE TITLE
progress bar may fail to update: busy cpu?

### ISSUE THREAD

**Issue by daviehh** [julia-vscode/julia-vscode#1769]
In the insider release, with https://github.com/julia-vscode/julia-vscode/pull/1579 the progress bar is supported. However, the progress bar may fail to update when there's no print outputs/cpu busy:

```julia
using ProgressLogging
using LinearAlgebra

function f1(nx)
    @withprogress name = "test" for i=1:nx
        m = rand(200, 200)
        ev = eigen(m)
        (i <= 10 || mod(i,20) == 0) && println("now: $i")
        @logprogress i / nx
    end
end

function f2(nx)
    @withprogress name = "test" for i=1:nx
        m = rand(200, 200)
        ev = eigen(m)
        (i <= 10 || mod(i,20) == 0) && println("now: $i")
        sleep(.01)
        @logprogress i / nx
    end
end
```

the `eig` there is to make the cpu busy in the loop. 

- Expected behavior:
running the code above in a REPL without vscode, both `f1(1000)` and `f2(1000)` updates the progress bar as expected with e.g TerminalLoggers.jl

- Actual behavior:
`f1(1000)` does not update progress bar correctly while `f2(1000)` works as expected. Incidentally,

```julia
function f3(nx)
    @withprogress name = "test" for i=1:nx
        m = rand(200, 200)
        ev = eigen(m)
        println("now: $i")
        @logprogress i / nx
    end
end
```
also makes the progress bar update as expected.

**Comment by daviehh** [comment]
more minium example:

```
function f1(nx)
    @withprogress name = "test" for i=1:nx
        m = rand(200, 200)
        ev = eigen(m)
        (i <= 10 || mod(i,20) == 0) && println("now: $i")
        # println(i)
        @logprogress i / nx
    end
end

function f2(nx)
    @withprogress name = "test" for i=1:nx
        m = rand(200, 200)
        ev = eigen(m)
        (i <= 10 || mod(i,20) == 0) && println("now: $i")
        sleep(.001)
        @logprogress i / nx
    end
end
```

`f1(1000)` does not update progress bar correctly while `f2(1000)` works as expected

**Comment by pfitzseb** [comment]
Ok, so the issue here is that we send the notifications in a different task; the evaluation task only fills a queue. Now, every `yield` in user code allows (roughly) one message to be sent, but that's obviously not enough as outlined by the example above.

Switching to an unbuffered implementation with e.g.
```diff
diff --git a/scripts/packages/VSCodeServer/src/progress.jl b/scripts/packages/VSCodeServer/s
rc/progress.jl
index d095672..97876d5 100644
--- a/scripts/packages/VSCodeServer/src/progress.jl
+++ b/scripts/packages/VSCodeServer/src/progress.jl
@@ -3,7 +3,9 @@ struct VSCodeLogger <: Logging.AbstractLogger end
 function Logging.handle_message(j::VSCodeLogger, level, message, _module,
                                 group, id, file, line; kwargs...)
     isprogress = try_process_progress(level, message, _module, group, id, file, line; kwargs...) do progress
-        JSONRPC.send_notification(conn_endpoint[], "repl/updateProgress", progress)
+        message = Dict("jsonrpc" => "2.0", "method" => "repl/updateProgress", "params" => progress)
+        message_json = JSON.json(message)
+        JSONRPC.write_transport_layer(conn_endpoint[].pipe_out, message_json)
     end isa Some
 
     if isprogress
```
fixes the issue, but might cause trouble down the line. Putting the write/read tasks on another thread should also fix the issue, I think.

@davidanthoff: Opinions on an unbuffered `send_notification`?

**Comment by aviatesk** [comment]
> the issue here is that we send the notifications in a different task; the evaluation task only fills a queue

I couldn't follow this; [`dispatch_msg`](https://github.com/julia-vscode/JSONRPC.jl/blob/a3f05f55fe9e2e78187a0f7e8d6030cec587a753/src/typed.jl#L57-L57) pushes a response only after `handler.func` (`repl_runcode_request`, this case) has been finished, and so the queue is empty while evaluation, no ?

**Comment by pfitzseb** [comment]
So [`send_notification` pushes the notification into the connection's `out_msg_queue`](https://github.com/julia-vscode/JSONRPC.jl/blob/a3f05f55fe9e2e78187a0f7e8d6030cec587a753/src/core.jl#L155), which is then [emptied in another task (started in `run`)](https://github.com/julia-vscode/JSONRPC.jl/blob/a3f05f55fe9e2e78187a0f7e8d6030cec587a753/src/core.jl#L92-L106).

If the user code never `yield`s that task will never do any work until the end, when it writes all progress messages at once. If the user code occasionally `yield`s, that task will write roughly one message per `yield`, which can be slower than messages are put into the queue. Putting a `yield` into `handle_message` seems like it should be enough, but also results in delayed messages.

**Comment by davidanthoff** [comment]
I think https://github.com/julia-vscode/JSONRPC.jl/issues/25 would solve this, right? I don't remember why I removed it after having added it at some point, but I think that would be the API that would be the most natural right now.

> Putting the write/read tasks on another thread should also fix the issue, I think.

In theory that would be the correct approach, but that won't work with the current depth-first scheduler in Julia. I think there were various discussions on the Julia repo about thread pools that each might have different schedulers, or various other solutions to this problem. In general having a task/thread system that is not depth-first for the kind of situation we have here would be incredibly helpful (also for the debugger, e.g.), but unless this lands in 1.6 (which I don't think it will), we probably have to find other ways here for the next couple of years, right? We essentially would need a LTS release that has a solution for this.

**Comment by pfitzseb** [comment]
That `flush` implementation does the trick, yes. I'll add that back in.

**Comment by pfitzseb** [comment]
https://github.com/JuliaLang/julia/issues/34267#issuecomment-573507670 might be worth looking into as well.
